core/async/src/test_loop/test_loop_sender.rs

```diff
@@ -0,0 +1,155 @@
+use std::any::type_name;
+use std::fmt::Debug;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::{Arc, Mutex, MutexGuard};
+
+use crate::futures::DelayedActionRunner;
+use crate::messaging::{Actor, CanSend, HandlerWithContext, MessageWithCallback};
+use crate::time::Duration;
+
+use super::delay_sender::DelaySender;
+use super::event_handler::LoopEventHandler;
+
+/// CallbackEvent for testloop is a simple event with a single callback which gets executed
+/// This is very versatile and we can potentially have anything as a callback. For example, for the case of Senders
+/// and Handlers, we can have a simple implementation of the CanSend function as a callback calling the Handle function
+/// of the actor.
+/// For DelayedActionRunner, we can have a callback that runs the function passed to run_later_boxed.
+pub struct CallbackEvent {
+    description: String,
+    callback: Box<dyn FnOnce() + Send>,
+}
+
+impl Debug for CallbackEvent {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("CallbackEvent").field(&self.description).finish()
+    }
+}
+
+/// Handler for CallbackEvent, include in testloop event enum to handle CallbackEvent
+pub fn callback_event_handler() -> LoopEventHandler<(), CallbackEvent> {
+    LoopEventHandler::new(|event: CallbackEvent, _| Ok((event.callback)()))
+}
+
+/// TestLoopSender implements the CanSend methods for an actor that can Handle them. This is similar to our pattern
+/// of having an ActixWarpper around an actor to send messages to it. For the purposes of testloop, we keep a copy of
+/// the DelaySender<CallbackEvent> that is used to schedule callbacks on the testloop to execute either the
+/// actor.handle() function or the DelayedActionRunner.run_later_boxed() function.
+pub struct TestLoopSender<A> {
+    actor: Arc<Mutex<A>>,
+    pending_events_sender: DelaySender<CallbackEvent>,
+    shutting_down: Arc<AtomicBool>,
+}
+
+// Since testloop is single-threaded, we can safely implement Send and Sync for TestLoopSender
+// This was specifically required as rust was complaining the traits Sync/Send are not implemented for actor A
+unsafe impl<A> Send for TestLoopSender<A> {}
```

Comment:
why is unsafe needed here?

Reply:
Unfortunately this rabbit hole is quite deep. This ties in with the question for why do we need Mutex in single thread application.

The CanSend interface requires us to have the object (in our case TestLoopSender<A>) implement the Sync+Send interface. This isn't directly required for test loop but the Actix implementations of PeerManagerActor and PeerActor require this.

For TestLoopSender<A> to implement Sync+Send, rust requires actor A to implement Send which isn't always guaranteed. In our case since we know what's going on and we have higher control over the components, it's fine to tell rust "just assume A implements Send" or rather "just assume TestLoopSender<A> implements Sync+Send".

There's also the AsyncComputationSpawner etc. etc. for which the most generic interface requires Send.

I hope this explanation makes sense. To fry one's brain, I'd recommend to explore commenting that unsafe impl block code and seeing what breaks :)